### ISSUE TITLE
SearchIO model pop has untested and potentially unused parameter default

### ISSUE THREAD

**Issue by fabianegli** [biopython/biopython#3604]
I am not quite sure what to make of the default parameter here:

https://github.com/biopython/biopython/blob/870f623005ed1b95df877f618cbb12a815bd5696/Bio/SearchIO/_model/query.py#L599-L666

I don't see the added value of adding this default. Am I missing something? 

A simple `popped = d.pop() if d else default` does the trick without all the extra object instantiation and testing.

I suggest removing the default parameter to make the pop behave as a pop on a builtin dict.

**Comment by bow** [comment]
Hi @fabianegli 

Good to see a suggestion here :). The reason for this implementation is actually to mimic the default `dict.pop` behavior, with a possibility of setting a default value if the value does not exist (as the docstring and comment says).

`d.pop("foo")` raises a KeyError if `foo` doesn't exist. This is a behavior that we would like to mirror in `QueryResult`.

However, I do also think it is handy to set a default (like in `getattr`), which will be the return value if the key does not exist. A custom `__marker` object was chosen, because this is the only way to detect whether a default has been set or not. `None` does not suffice, as it can also be used as the given default value.

So in short, adding a `default` this way is to make the `pop` function more concise (ala `getattr`) but still keep a default behavior that is the same as `dict.pop`.

**Comment by peterjc** [comment]
Doesn't the Python dict's pop have a default argument?

```
pop(...) method of builtins.dict instance
    D.pop(k[,d]) -> v, remove specified key and return the corresponding value.
    If key is not found, d is returned if given, otherwise KeyError is raised
```

**Comment by bow** [comment]
> Doesn't the Python dict's pop have a default argument?
> 
> ```
> pop(...) method of builtins.dict instance
>     D.pop(k[,d]) -> v, remove specified key and return the corresponding value.
>     If key is not found, d is returned if given, otherwise KeyError is raised
> ```

Actually, it does indeed. Not documented explicitly in the docstring, but indeed it can be set:

```
>>> {"a": 1}.pop("b")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'b'
>>> {"a": 1}.pop("b", 100)
100
>>> 
```

So yes, this is actually then to mimic the stdlib `.pop` method :).

**Comment by fabianegli** [comment]
@bow My example in the question is indeed not very useful. I am glad the discussion still developed in the direction I hoped it would. 

 I now see the beauty in using an object as a "someting" that can reasonably be expected to never be the default set by a user.

After looking a little bit longer at the code I think there are two code cleanups that might be considered:

1. Remove the default parameter in the pop function in line 661. If the key is in the dict, which the preceding if tests for, the default will never be needed. Also, at this point, if it were needed, it could as well be the __marker object which should never be returned.
2. Unindent lines 662 to 664. At this point a return has been hit if the key exists. This would put all the default related handling on the same level.

**Comment by bow** [comment]
@fabianegli That sounds good :). I'll respond to your points below.

> Remove the default parameter in the pop function in line 661. If the key is in the dict, which the preceding if tests for, the default will never be needed. Also, at this point, if it were needed, it could as well be the __marker object which should never be returned.

Yes ~ indeed if the hit ID value resolves to an alternative hit ID, we know that the `self.pop` would return a value and therefore `default` is essentially uneeded.

> Unindent lines 662 to 664. At this point a return has been hit if the key exists. This would put all the default related handling on the same level.

This breaks the function, right? I'm assuming you meant unindenting it out of the `KeyError` except block. We only want to raise `KeyError` if the following conditions are met:

* The hit ID is not a primary ID
* The hit ID is not an alternative ID
* No default value is ever supplied

Which is checkable only at that indentation level, right?

**Comment by fabianegli** [comment]
@bow Indeed, I meant unindenting one level to take the if and its content out of the exception block. The first two conditions in your condition list result in a return from the method - so these conditions are met after the try except block -  and the last condition is being checked by the if statement itself. So taking it out from the try/except block makes sense to me. For me it also makes sense to have the check if a default has been set on the same level as the `return default` because this is the level where default is going to be used.

**Comment by bow** [comment]
> So taking it out from the try/except block makes sense to me. For me it also makes sense to have the check if a default has been set on the same level as the return default because this is the level where default is going to be used.

Hmm..I don't completely agree that the level in which `default` is used justifies a de-indent. It makes more sense for `default is self.__marker` check to be done within a `KeyError` clause, as this check is done in the context of a key not found, in spite of having the same effect either way.

There are three places where `KeyError` is used in that `try ... except` block, all for different yet related scenarios. Having them in one block tells the reader that the `try ... except` block handles various cases involving missing keys. De-indenting the `if`-block does not provide any tangible (e.g. performance) benefits, but IMO decreases readability. In fact, looking at this again, would it not make more sense to put `return default` inside the except block? So indenting that, instead of de-indenting the previous `if`-clause?

**Comment by fabianegli** [comment]
I agree, that the different discussed placements will not result in a different behaviour.

For me, a different pattern is more important here: the early return. Once a certain condition is met that results in a return, we have the luxury to forget about it. That is the reason why early returns are useful to reduce mental load by removing step by step the amount of information needed to make the next decision. I think this naturally translates into fewer indentation levels in the code.

I see a different inconsistency in the code, too. For the standard hit id dict a try except pattern is chosen in contrast to the key existence check for the __alt_hit_ids dict. 

We could also put the raise at the end to have an even more consistent linearly logic flow like this:

```python
        if hit_key in self._items:
            hit = self._items.pop(hit_key)
            # remove all alternative IDs of the popped hit
            for alt_id in hit.id_all[1:]:
                try:
                    del self.__alt_hit_ids[alt_id]
                except KeyError:
                    pass
            return hit

        if hit_key in self.__alt_hit_ids:
            return self.pop(self.__alt_hit_ids[hit_key])

        # hit_key doesn't exist
        # return the default if one has been set
        if default is not self.__marker:
            return default

        # no valid return value for hit_key exists.
        raise KeyError(hit_key) from None
```

I know about the try mantra, I just find the above code more readable.

If you still think that the indentation should remain, I'll revert it.

What do you think about moving the raise to the end of the function? I think the comment line discribes well why I put it at the end (try to make it work until we know it doesn't).

**Comment by fabianegli** [comment]
We could indeed also indent the `return default` line. 

Here for comparison with complete adoption of the try except pattern:

```python
        try:
            hit = self._items.pop(hit_key)
            # remove all alternative IDs of the popped hit
            for alt_id in hit.id_all[1:]:
                try:
                    del self.__alt_hit_ids[alt_id]
                except KeyError:
                    pass
            return hit
        except KeyError:
            try:
                return self.pop(self.__alt_hit_ids[hit_key])
            except KeyError:
                # hit_key doesn't exist
                # return the default if one has been set
                if default is not self.__marker:
                    return default
                # no valid return value for hit_key exists
                raise KeyError(hit_key) from None
```


An one version with a single return:

```python
        try:
            hit = self._items.pop(hit_key)
            # remove all alternative IDs of the popped hit
            for alt_id in hit.id_all[1:]:
                try:
                    del self.__alt_hit_ids[alt_id]
                except KeyError:
                    pass
        except KeyError:
            try:
                hit = self.pop(self.__alt_hit_ids[hit_key])
            except KeyError:
                # hit_key doesn't exist
                # use the default if one has been set
                if default is not self.__marker:
                    hit = default
                else:
                    # no valid return value for hit_key exists
                    raise KeyError(hit_key) from None
        return hit
```

**Comment by fabianegli** [comment]
The above code version have been run successfully against the tests including the new test in the PR branch.

**Comment by bow** [comment]
@fabianegli A lot of good points raised here :+1: ~ For this point in particular:

> For me, a different pattern is more important here: the early return. Once a certain condition is met that results in a return, we have the luxury to forget about it. That is the reason why early returns are useful to reduce mental load by removing step by step the amount of information needed to make the next decision. I think this naturally translates into fewer indentation levels in the code.

Yes, here I agree. I find early returns also useful in 'pruning' the context as we traverse the executions line-by-line.

I do have to admit, looking at the code again after all these years have made me feel that it could have been written differently (i.e. the current version in `master` has room for improvements). So I'm glad this is being discussed now :).

For the options you are proposing in particular, I do feel partial towards the last option (i.e. the single-return version). Here the reader can also see the `try..except` block as the part where `hit` is resolved, and we only re-raise the exception (the very last `raise KeyError`) if all options have been exhausted.

One more suggestion here, is to actually use [`contextlib.suppress`](https://docs.python.org/3/library/contextlib.html#contextlib.suppress):

```python
from contextlib import suppress

        try:
            hit = self._items.pop(hit_key)
            # remove all alternative IDs of the popped hit
            for alt_id in hit.id_all[1:]:
                with suppress(KeyError):
                    del self.__alt_hit_ids[alt_id]
         except KeyError:
            try:
                hit = self.pop(self.__alt_hit_ids[hit_key])
            except KeyError:
                # hit_key doesn't exist
                # use the default if one has been set
                if default is not self.__marker:
                    hit = default
                else:
                    # no valid return value for hit_key exists
                    raise KeyError(hit_key) from None
        return hit
```

This gives also a more explicit distinction of the way exceptions are handled: `try..except` blocks handles cases where we do want to actively handle an exception, while `suppress` really means that we want to completely ignore the exception.

@peterjc Would it be ok here to use the `contextlib.suppress` statement? This is different from other parts in Biopython, where `try..except..pass` is actually used to ignore exceptions, though I do think `contextlib.suppress` is one of those things in the stdlib that deserves more use :).

**Comment by peterjc** [comment]
I would lean towards the traditional try/except and raise as needed - but that is in part from lack of familiarity with ``contextlib.suppress``,  in this example I would initially be wondering what it was doing.

**Comment by fabianegli** [comment]
We could in that place also use the fabulous `pop` with a default

```python
            for alt_id in hit.id_all[1:]:
                self.__alt_hit_ids.pop(alt_id, None)
```

**Comment by bow** [comment]
@fabianegli That indeed is an option :). I'd be fine with having that instead :+1: ~
